### DISCUSSION TITLE
Ramda ts proposal

### DISCUSSION THREAD

**Discussion by Harris-Miller** [ramda/ramda#3329] [Ideas]
Hi Ramda team.

As you may have noticed, made a lot of PRs to the `ts` branch in the past week. @kedashoe sent me this e-mail in response

> Hey Harris
> Thanks for all the types PRs. Honestly it is too much for me to review! I'm not a typescript expert in any event.
> I'd like to propose I create a new repository, @ramda/types, and we move all the type work there. I'll make you a maintainer. There seemed to be some excitement for getting this stuff into @ramda/ramda, but pretty much everyone who was advocating for it has ghosted once I started working on it.
> I would say it is more difficult to see a way forward there with the current approach vs working in a separate repo where we can much more easily/visibly release and have people actually use it. At some point in the future maybe we could merge it back into @ramda/ramda

I'm happy to take on that ownership and responsibility, and wanted to use this discussion to go over that option and an alternative

As someone who is personally a huge fan of `ramda` and works 90% in typescript both professionally and personally, the current state of types for `ramda` is the biggest hurdle for me to get others to adopt over say `lodash`, and is why I've begun contributing with my recent PRs

### Current typings
The community-driven `@types/ramda` that lives in the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/ramda) repo works fine currently but is not ideal to keep. 

Mostly because of the separation of ownership between the two. `@types/ramda` is also currently inaccurate in a lot of places, missing types for curried varieties, or straight up missing types for some functions. The block-comments are also copy-paste'd from the core repo and would have to be updated manually and separately and manually. None of that is ideal.

As typescript becomes more ubiquitous in the javascript ecosystem, projects written directly in typescript or have in-repo-defined `.d.ts` files are becoming more the norm. And for the longevity of `@ramda/ramda`, this is a pattern that should be adopted. Clearly, with the `ts` branch, this process is already underway.

### Separate `@ramda/types` repo
@kedashoe's proposal is that a `@ramda/types` repo is created to solely house the typings for Ramda. The way I see this working is by simply moving everything that is new from the `ts` branch over to the new repo.

The idea here is not to replace `@types/ramda` with `@ramda/types`, but rather that `@ramda/ramda` has the new `@ramda/types` as a `devDependency`, and _build_ the types as part of the release pipeline. This way, the types are part of the npm packages, rather than a separate dependency.`@ramda/types` would never be consumed independently from `@ramda/ramda`.

There are advantages and disadvantages to this strategy. As stated in the e-mail above, the main reason for this separation is repo ownership. The core `@ramda/ramda` repo will continue to be owned by the core team, while the `@ramda/types` repo would be owned by myself and any additional core contributors.

The upside is a separation of PRs, and other GitHub features like discussions, projects, wiki, etc. The downside would be getting `@ramda/types` into `@ramda/ramda`. Having these be two separate repos adds a level of complexity that may be more harmful than helpful.

### Codeowners in core repo
Alternatively we could go the [codeowners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) route. This essentially keeps everything in the core `@ramda/ramda` repo and delegates certain files/folders to separate owners for PR approval. This solves the ownership issue @kedashoe pointed out in his e-mail.

A GitFlow style of development would continue with the `ts` branch getting merged into `master` after reaching specific milestones

## Development and Release
Regardless of which strategy is chosen, development and release remain the same
* a "development" branch is used for PR acceptance over the course of a milestone
* the "development" branch is merged into the "release" branch to trigger a new release

For `@ramda/types`, this means `development` -> `master`, then a PR to bump `@ramda/types` in `@ramda/ramda`.

For Codeowners, this means `ts` -> `master`

## Milestones

The `ts` branch started out mostly as a copy of `@types/ramda`. Which is fine, but as I mentioned earlier, there are a lot problems with the state of `@types/ramda`, so doing an upfront release of `ramda` with built-in types would end up with possible Breaking changes down the line to types. So ideally, getting _all_ the types in-repo and done correctly would be ideal for an initial release.

That however, may take a lot of time, pushing back the release for "Ramda w/ types" indefinitely. IMHO releasing an MVP version of "Ramda w/ types", with any breaking changes up-front, would be acceptable. Post-MVP would be a series of non-breaking, progressive releases until typing for Ramda is at 100%

I have a lot of experience doing this professionally and is something I can spec out in detail.

## Final Thoughts
Personally I think the Codeowners strategy is the way to go. Having a separate repo is likely going to be more of a pain than it's worth. And Codeowner files are very simple to maintain

Thank you to the Ramda core team for the opportunity to help release typings in an official capacity!

**Comment by Harris-Miller** [comment]
@kedashoe bump

**Comment by kedashoe** [comment]
Spoke to @Harris-Miller via email, we are going to try a separate 'ramda/types' repo for now.

**Comment by CrossEye** [comment]
I have had no time to dedicate to Ramda for several months, and won't for another two weeks or so, but after that, I expect to be back, and pushing toward `1.0`.  I think this decision is great.  @kedashoe, I think you have the authority to create a new repo and team for it.  If not, please let me know.  @Harris-Miller: Thanks for al the work you've been doing lately.

**Reply by Harris-Miller** [reply]
@kedashoe @CrossEye I've been super busy lately, after Thanksgiving break I'll be able to dedicate some time to getting the new private `@ramda/types` repo up and running and work out a game plan

**Comment by Harris-Miller** [comment]
@kedashoe @CrossEye Ok I'm back. Here's the plan.

* `@ramda/types` is going to be simply the `ts` branch of `@ramda/ramda`, but with just the type additions.
* This way we can continue to merge my changes along with others in the community in a repo we control
* The testing strategy I'm going to change to be the same style of testing used by DefinitelyTyped
  * This is because we can't have a circular dependency between `@ramda/types` and `@ramda/ramda`, and the current testing strategy requires actual concrete functions, while the `@types/ramda` method does not.
* Once `@ramda/types` is set up and published on npm, I will make an MR in [@types/ramda](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/ramda) that will remove everything that is currently there, and instead just re-export `@ramda/types`
  * the goal with that we move _where_ the types lives from a community owned repo to one we own. But at the same time, continue to use `@types/ramda` as the way to using `ramda` with typings
* Once `@ramda/types` is fully up to date with `ramda@0.28.0`, we'll look into migrating `@ramda/types` back into `@ramda/ramda`, and deprecating `@types/ramda`
  * This is preferable in the long run. It will make long-term type maintainability easier, and any signature changes to functions will include updates to the types in the same MR

I'm working on standing up `@ramda/types` now and will push something up later this week. At that point I will make it `public` and blast out for community support

Let me know what you guys think about this plan

**Reply by kedashoe** [reply]
I'm not sure about the MR to DefinitelyTyped at what would seem to be a bit of an early stage, but yall know more about how they operate than I do so if you think that is best it is fine with me. Everything else sounds good and we can talk about the bullet point more as they come up.

**Reply by Harris-Miller** [reply]
The MR to DefinitelyTyped is about single-source-of-truth. The first release of `@ramda/types` should be a one-to-one of `@types/ramda`. That way we don't have the community contributing to `@types/ramda`, and instead just to `@ramda/types` directly.

And with `@types/ramda` just re-exporting `@ramda/types`, we can get the updated type definitions out into the world much faster. IMHO I think it would be good to get the updated types out there quicker for purposes of both feedback and community-driven corrections

**Reply by adispring** [reply]
@Harris-Miller First, thanks for your contribute. There are some hard but important problems that need to solve for ramda working friendly with ts, I list some here:

1. how to do check and reason composed functions' type easily and correctly within high order functions: such as compose, pipe, converge, useWith, evolve, applySpec, and so on.
2. how ts support curry friendly. when compose curried functions in high order functions, curry make things terrible.

Do you have any plan or thought about how to resolve the above problems?

**Reply by Harris-Miller** [reply]
### `compose`/`pipe`

I assume you mean not having to type out an overload for each arity of `compose`/`pipe`. I don't think there is a better solution. Typescript types in general are not very FP friendly. And while it is indeed extremely powerful, it doesn't have the ability to generate stuff like this. [ts-toolbelt](https://github.com/millsp/ts-toolbelt/blob/master/sources/Function/Compose/Multi/Sync.ts) overloads in the same way that [@types/ramda ](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ramda/index.d.ts#L782) currently does. This is simply a limitation of typescript

### curring

Currying on the other hand can be generated, as seen with [ts-toolbelt](https://github.com/millsp/ts-toolbelt/blob/master/sources/Function/Curry.ts), but not without caveats. While those do _work_, IMHO writing out our own overloaded types for each function yield a much greater user experience. They are also better when it comes to deriving types in other functions like `compose` and `pipe`. A good example of this would be how I typed out the [all function](https://github.com/Harris-Miller/ramda/blob/ts-all/types/all.d.ts)

It's a lot of worth, but work I'm willing to take on. Additionally, give that 90% of ramda functions aren't changing anytime soon, neither will the types. Once they are completed, we will almost never have to touch them again

**Comment by Harris-Miller** [comment]
# Update

At the time of making the issue, I confirmed that what I have in `@ramda/types` is in parity with `@types/ramda` (see [here](https://github.com/ramda/types/issues/1)). This means that I can effectively re-export `@ramda/types` from `@types/ramda` (see my first post in this thread for an explanation)

At this point, I'm ready to make the repo public and do that. I will need some help setting up npm publishing. Not sure if ramda owns the `@ramda` namespace on `npm` or not, or if you even want to go that route. The name of the types package doesn't matter. It just needs to be deployed and re-exported by `@types/ramda`. (It's all temporary anyhow since the eventual goal is to re-integrate into the core ramda package).

I am also working on a [Master List](https://github.com/ramda/types/issues/2) which defines what work is needed for each function's type definitions. Once that is done I'll split it out into manageable chunks, prioritize, and start reaching out for community help.

Let me know if you have any questions. Thank you!

**Comment by Harris-Miller** [comment]
Kevin and Scott,

I was hoping we could make ramda/types <https://github.com/ramda/types>
public now and your help with a few items
First
a) Get publish pipelines set up for it
b) Get an MR set up in @types/ramda to re-export ramda/types
c) Profit!

At minimum, I would like to get the ramda/types repo made public so I can
manage the work ahead via Github issues/projects, etc, and work can at
least begin in the DefinatelyTyped repo until we're ready to full migrate
things over to ramda/types

I'm very much invested in getting Ramda typings to where they should be.
Typescript adoption has grown considerably over the past few years and at
my company the lack of good ramda types is the major blocker for full
adaption. I've heard elsewhere that that is the main case as well. One of
my roles as a Technical Lead at my company is to help drive innovation, and
that blocker for adaption has been challenging. Ramda works *very* well
with modern web development, especially in React-land where everything is
immutable. I can't tell you how many bugs I have solved by pointing to my
junior devs simply by introducing them to ramda to use over lodash.
First-class typings, IMHO, will help excel Ramda adoption by the community
at large

Thank you!

On Tue, Nov 29, 2022 at 6:30 PM wangzengdi ***@***.***> wrote:

> @Harris-Miller <https://github.com/Harris-Miller> First, thanks for your
> contribute. There are some hard but important problems that need to solve
> for ramda working friendly with ts, I list some here:
>
>    1. how to do check and reason composed functions' type easily and
>    correctly in pipeline (compose functions with compose/pipe) .
>    2. how ts support curry friendly. when compose curried functions in
>    pipeline, curry make things terrible.
>
> Do you have any plan or thought about how to resolve the above problems?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ramda/ramda/discussions/3329#discussioncomment-4269529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVCTK5TNQTFHLSHQNGB6HLWK2U4PANCNFSM6AAAAAARMOLT5I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Harris Miller
301-404-7090
***@***.***

**Reply by kedashoe** [reply]
The repo is now public! I can help set up the public action as well

**Comment by Harris-Miller** [comment]
I'll start another discussion in that repo for it. Thank you!

On Thu, Mar 9, 2023 at 4:12 PM Kevin Wallace ***@***.***>
wrote:

> The repo is now public! I can help set up the public action as well
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ramda/ramda/discussions/3329#discussioncomment-5260540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVCTKYBYFTR5JNAWNNQBOLW3JPVRANCNFSM6AAAAAARMOLT5I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Harris Miller
301-404-7090
***@***.***

**Comment by Harris-Miller** [comment]
Side question: any thoughts on releasing `0.29.0`? `master` is currently 55
commits ahead of `v0.28.0`, and there are a lot of great bug fixes and at
least one new function (`isNotNil`) in there

On Thu, Mar 9, 2023 at 6:20 PM Harris Miller <
***@***.***> wrote:

> I'll start another discussion in that repo for it. Thank you!
>
> On Thu, Mar 9, 2023 at 4:12 PM Kevin Wallace ***@***.***>
> wrote:
>
>> The repo is now public! I can help set up the public action as well
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ramda/ramda/discussions/3329#discussioncomment-5260540>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAVCTKYBYFTR5JNAWNNQBOLW3JPVRANCNFSM6AAAAAARMOLT5I>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Harris Miller
> 301-404-7090
> ***@***.***
>


-- 
Harris Miller
301-404-7090
***@***.***
